src/gc/gc.cpp

```diff
@@ -34211,12 +34085,91 @@ HRESULT GCHeap::Initialize()
     uint32_t nhp_from_config = 0;
 
 #ifdef MULTIPLE_HEAPS
+    AffinitySet config_affinity_set;
+
+    // Get the affinity set configured by the user
+    uintptr_t heap_affinity_mask = GCConfig::GetGCHeapAffinitizeMask();
+    if (heap_affinity_mask != 0)
+    {
+        for (size_t i = 0; i < 8 * sizeof(uintptr_t); i++)
+        {
+            if (heap_affinity_mask & ((uintptr_t)1 << i))
+            {
+                config_affinity_set.Add(i);
+            }
+        }
+    }
+    else
+    {
+        GCConfigStringHolder cpu_index_ranges_holder(GCConfig::GetGCHeapAffinitizeRanges());
+        const char* cpu_index_ranges = cpu_index_ranges_holder.Get();
+
+        // The cpu index ranges is a comma separated list of indices or ranges of indices (e.g. 1-5).
+        // Example 1,3,5,7-9,12
+
+        if (cpu_index_ranges != NULL)
+        {
+            char* number_end;
+
+            do
+            {
+                size_t start_index = strtoul(cpu_index_ranges, &number_end, 10);
+
+                if (number_end == cpu_index_ranges)
+                {
+                    // No number found, invalid format
+                    break;
+                }
+
+                size_t end_index = start_index;
+
+                if (*number_end == '-')
+                {
+                    char* range_end_start = number_end + 1;
+                    end_index = strtoul(range_end_start, &number_end, 10);
+                    if (number_end == range_end_start)
+                    {
+                        // No number found, invalid format
+                        break;
+                    }
+                }
+
+                if ((start_index < MAX_SUPPORTED_CPUS) && end_index < (MAX_SUPPORTED_CPUS))
+                {
+                    for (size_t i = start_index; i <= end_index; i++)
```

Comment:
> start_index; i <= end_index; i++) [](start = 35, length = 34)

since we are checking for errors here, it would also be an error to have an ill formatted string here (eg, end_index is < start_index; '-' not followed by a number). do we want to also return CLR_E_GC_BAD_AFFINITY_CONFIG for those cases?

Reply:
I was thinking about how strict we should with failing on parse errors in this string. But I guess that you are right, since otherwise the user would not be able to tell when he used a wrong format.